### PR TITLE
refactor(drawer): rewrite with hook

### PR DESCRIPTION
**Pull request by fardad-dev**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 💡 Background and solution

https://github.com/ant-design/ant-design/projects/6

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Refactor Drawer with react hooks. |
| 🇨🇳 Chinese | 使用 react hooks 重构 Drawer。 |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/drawer/index.tsx [hunk 1]
```diff
@@ -61,271 +66,270 @@ interface InternalDrawerProps extends DrawerProps {
...
-
-  public componentDidUpdate(preProps: DrawerProps) {
-    const { visible } = this.props;
-    if (preProps.visible !== visible && this.parentDrawer) {
-      if (visible) {
-        this.parentDrawer.push();
-      } else {
-        this.parentDrawer.pull();
+const defaultCloseIcon = <CloseOutlined />;
+// React.Component<InternalDrawerProps, IDrawerState>
```
Comment (hengkx) [inline components/drawer/index.tsx:70]: removed

File: components/drawer/index.tsx [hunk 2]
```diff
@@ -61,271 +66,270 @@ interface InternalDrawerProps extends DrawerProps {
...
+      title,
+      headerStyle,
+      onClose,
+      footer,
+      footerStyle,
+      ...rest
+    },
+    ref,
+  ) => {
+    const [, forceUpdate] = React.useState({});
```
Comment (hengkx) [inline components/drawer/index.tsx:103]: https://github.com/ant-design/ant-design/blob/master/components/_util/hooks/useForceUpdate.ts

File: components/drawer/index.tsx (hunk 1 above)
Comment (hosseinmd) [reply components/drawer/index.tsx:71]: I don't understand, Could you explain what can i do?

File: components/drawer/index.tsx [hunk 3]
```diff
@@ -61,271 +67,270 @@ interface InternalDrawerProps extends DrawerProps {
...
-
-  public componentDidUpdate(preProps: DrawerProps) {
-    const { visible } = this.props;
-    if (preProps.visible !== visible && this.parentDrawer) {
-      if (visible) {
-        this.parentDrawer.push();
-      } else {
-        this.parentDrawer.pull();
+const defaultCloseIcon = <CloseOutlined />;
+// React.Component<InternalDrawerProps, IDrawerState>
```
Comment (hengkx) [inline components/drawer/index.tsx:71]:
```suggestion

```

**Comment by hosseinmd** [comment]
@hengkx Done.

**Comment by hengkx** [comment]
没有改变的文件，建议不要 产生改变 记录

**Comment by yoyo837** [comment]
文件变化没带进来勉强可以 Squash 接受，没记录固然更好。

**Comment by hengkx** [comment]
文件变化 应该是带进来了

**Comment by yoyo837** [comment]
> 文件变化 应该是带进来了

那不行，重新rebase.

**Comment by hosseinmd** [comment]
I don't know Chinese. Could you please speak English.

**Comment by yoyo837** [comment]
> I don't know Chinese. Could you please speak English.

Sorry for that, please rebase again. Commits should be clean.

**Comment by hosseinmd** [comment]
Thanks, I did that.

**Comment by afc163** [comment]
Please push with `--force` after rebase master.

File: components/drawer/index.tsx [hunk 4]
```diff
@@ -61,271 +67,269 @@ interface InternalDrawerProps extends DrawerProps {
...
+      width = 256,
+      height = 256,
+      closable = true,
+      placement = 'right' as placementType,
+      maskClosable = true,
+      mask = true,
+      level = null,
+      keyboard = true,
+      push = defaultPushState,
+      closeIcon = defaultCloseIcon,
```
Comment (hengkx) [inline components/drawer/index.tsx:83]:
```suggestion
      closeIcon = <CloseOutlined />,
```

File: components/drawer/index.tsx [hunk 5]
```diff
@@ -61,271 +67,269 @@ interface InternalDrawerProps extends DrawerProps {
...
-  }
-
-  public componentDidUpdate(preProps: DrawerProps) {
-    const { visible } = this.props;
-    if (preProps.visible !== visible && this.parentDrawer) {
-      if (visible) {
-        this.parentDrawer.push();
-      } else {
-        this.parentDrawer.pull();
+const defaultCloseIcon = <CloseOutlined />;
```
Comment (hengkx) [inline components/drawer/index.tsx:70]:
```suggestion

```

File: components/drawer/index.tsx [hunk 6]
```diff
@@ -61,271 +67,269 @@ interface InternalDrawerProps extends DrawerProps {
...
-      distance = push!.distance;
-    }
-    return parseFloat(String(distance || 0));
-  };
+    const isDestroyOnClose = destroyOnClose && !visible;
 
-  // get drawer push width or height
-  getPushTransform = (placement?: placementType) => {
-    const distance = this.getPushDistance();
+    const onDestroyTransitionEnd = React.useCallback(() => {
```
Comment (shaodahong) [inline components/drawer/index.tsx:153]: Why use `useCallback`?

Comment (hosseinmd) [reply components/drawer/index.tsx]: That was a mistake, I resolved that.
thanks

**Comment by hengkx** [comment]
conflict

**Pull request closed without merging**